### PR TITLE
fix bug: dpdk场景下添加删除ProviderNetwork失败

### PR DESCRIPTION
**Pull request by wangwangyusur288**
# Pull Request

## What type of this PR

dpdk场景下添加删除ProviderNetwork失败

- Bug fixes

修改在--with-hybrid-dpdk场景下，创建删除ProviderNetwork出现的问题。

### CODE REVIEW

File: pkg/daemon/init.go [hunk 1]
```diff
@@ -159,42 +159,50 @@ func (c *Controller) ovsCleanProviderNetwork(provider string) error {
...
-				continue
-			}
-			klog.V(3).Infof("removing ovs port %s from bridge %s", port, brName)
-			if err = c.removeProviderNic(port, brName); err != nil {
-				errMsg := fmt.Errorf("failed to remove port %s from external bridge %s: %w", port, brName, err)
-				klog.Error(errMsg)
-				return errMsg
+	if !isUserspaceDP {
+		// get host nic
+		if output, err = ovs.Exec("list-ports", brName); err != nil {
```
Comment (bobz965) [inline pkg/daemon/init.go:170]: need log  near err

File: pkg/daemon/init.go [hunk 2]
```diff
@@ -159,42 +159,50 @@ func (c *Controller) ovsCleanProviderNetwork(provider string) error {
...
+			return fmt.Errorf("failed to list ports of OVS bridge %s, %w: %q", brName, err, output)
+		}
+
+		// remove host nic from the external bridge
+		if output != "" {
+			for _, port := range strings.Split(output, "\n") {
+				// patch port created by ovn-controller has an external ID ovn-localnet-port=localnet.<SUBNET>
+				if output, err = ovs.Exec("--data=bare", "--no-heading", "--columns=_uuid", "find", "port", "name="+port, `external-ids:ovn-localnet-port!=""`); err != nil {
+					return fmt.Errorf("failed to find ovs port %s, %w: %q", port, err, output)
```
Comment (bobz965) [inline pkg/daemon/init.go:179]: need log near err

File: pkg/daemon/init.go [hunk 3]
```diff
@@ -159,42 +159,50 @@ func (c *Controller) ovsCleanProviderNetwork(provider string) error {
...
+				}
+				if output != "" {
+					continue
+				}
+				klog.V(3).Infof("removing ovs port %s from bridge %s", port, brName)
```
Comment (bobz965) [inline pkg/daemon/init.go:184]: klog.V(3) --> klog is better, it just log once.

File: pkg/daemon/init.go [hunk 4]
```diff
@@ -159,42 +159,50 @@ func (c *Controller) ovsCleanProviderNetwork(provider string) error {
...
+				if err = c.removeProviderNic(port, brName); err != nil {
+					errMsg := fmt.Errorf("failed to remove port %s from external bridge %s: %w", port, brName, err)
```
Comment (bobz965) [inline pkg/daemon/init.go:186]:
```suggestion
					err := fmt.Errorf("failed to remove port %s from external bridge %s: %w", port, brName, err)
```

File: pkg/daemon/init.go [hunk 5]
```diff
@@ -159,42 +159,50 @@ func (c *Controller) ovsCleanProviderNetwork(provider string) error {
...
 
-	// remove OVS bridge
-	klog.Infof("delete external bridge %s", brName)
-	if output, err = ovs.Exec(ovs.IfExists, "del-br", brName); err != nil {
-		return fmt.Errorf("failed to remove OVS bridge %s, %w: %q", brName, err, output)
-	}
-	klog.V(3).Infof("ovs bridge %s has been deleted", brName)
+		// remove OVS bridge
+		klog.Infof("delete external bridge %s", brName)
+		if output, err = ovs.Exec(ovs.IfExists, "del-br", brName); err != nil {
```
Comment (bobz965) [inline pkg/daemon/init.go:196]: log near err

File: pkg/daemon/ovs_linux.go [hunk 6]
```diff
@@ -1341,16 +1341,27 @@ func (c *Controller) transferAddrsAndRoutes(nicName, brName string, delNonExiste
...
+		return 0, err
 	}
 
-	if _, err = ovs.Exec(ovs.MayExist, "add-port", brName, nicName,
-		"--", "set", "port", nicName, "trunks="+strings.Join(trunks, ","), "external_ids:vendor="+util.CniTypeName); err != nil {
-		return 0, fmt.Errorf("failed to add %s to OVS bridge %s: %w", nicName, brName, err)
+	var mtu int
+	if !isUserspaceDP {
+		mtu, err = c.transferAddrsAndRoutes(nicName, brName, false)
+		if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ovs_linux.go:1353]: need log near err

File: pkg/daemon/ovs_linux.go [hunk 7]
```diff
@@ -1341,16 +1341,27 @@ func (c *Controller) transferAddrsAndRoutes(nicName, brName string, delNonExiste
...
+			return 0, fmt.Errorf("failed to transfer addresses and routes from %s to %s: %w", nicName, brName, err)
+		}
+
+		if _, err = ovs.Exec(ovs.MayExist, "add-port", brName, nicName,
+			"--", "set", "port", nicName, "trunks="+strings.Join(trunks, ","), "external_ids:vendor="+util.CniTypeName); err != nil {
+			return 0, fmt.Errorf("failed to add %s to OVS bridge %s: %w", nicName, brName, err)
```
Comment (bobz965) [inline pkg/daemon/ovs_linux.go:1359]: need log near err

Comment (wangwangyusur288) [reply pkg/daemon/ovs_linux.go]: > need log near err

Fixed, thank you.

**Pull request merged**